components/connectionmanager/src/eviction.rs

```diff
@@ -0,0 +1,818 @@
+/// _2) A peer may hold a rank as a multiple of 0.5, due to tie breaks splitting the rank._
+#[derive(Default, Clone, Copy, PartialEq, Debug)]
+pub struct EvictionRanks {
+    ip_prefix_bucket: f64,
+    time_connected: f64,
+    last_ping_duration: f64,
+    last_block_transfer: f64,
+    last_tx_transfer: f64,
+}
+
+impl EvictionRanks {
+    /// Returns and defines an array of all ranks that are not latency based.
+    ///
+    /// _**Includes:** [`Self::time_connected`], and [`Self::ip_prefix_bucket`]._
+    ///
+    /// _**Excludes:** [`Self::last_ping_duration`], [`Self::last_block_transfer`], and [`Self::last_tx_transfer`]._
+    #[inline]
+    fn none_latency_ranks(&self) -> [f64; 2] {
+        [self.time_connected, self.ip_prefix_bucket]
+    }
+
+    /// Returns and defines an array of all ranks in the struct.
+    ///
+    /// **Includes:** [`Self::last_ping_duration`], [`Self::time_connected`], [`Self::last_block_transfer`], [`Self::last_tx_transfer`], and [`Self::ip_prefix_bucket`].
+    #[inline]
+    fn all_ranks(&self) -> [f64; 5] {
+        [self.last_ping_duration, self.time_connected, self.last_block_transfer, self.last_tx_transfer, self.ip_prefix_bucket]
+    }
+
+    /// Returns the lowest rank of all ranks in the struct.
+    ///
+    /// Refer to [`Self::all_ranks`] for the ranks.
+    ///
+    /// _**Note**: The lowest rank is the "best" rank, as ranks are organized from low ("good") to high ("bad")._
+    ///
+    /// _**Includes:** [`Self::last_ping_duration`], [`Self::time_connected`], [`Self::last_block_transfer`], [`Self::last_tx_transfer`], and [`Self::ip_prefix_bucket`]._
+    #[inline]
+    pub fn lowest_rank(&self) -> f64 {
+        self.all_ranks().into_iter().reduce(move |a, b| a.min(b)).expect("expected at least one rank")
+    }
+
+    /// Returns the highest rank of the non-latency ranks / variables.
+    ///
+    /// _**Note**: The highest rank is the "worst" rank, as ranks are organized from low ("good") to high ("bad")._
+    ///
+    /// _**Includes:** [`Self::time_connected`], and [`Self::ip_prefix_bucket`]._
+    ///
+    /// _**Excludes:** [`Self::last_ping_duration`], [`Self::last_block_transfer`], and [`Self::last_tx_transfer`]._
+    #[inline]
+    pub fn highest_none_latency_rank(&self) -> f64 {
+        self.none_latency_ranks().into_iter().reduce(move |a, b| a.max(b)).expect("expected at least one rank")
+    }
+}
+
+pub type EvictionIter<'a> = Box<dyn Iterator<Item = (&'a Peer, EvictionRanks)> + 'a>;
+pub type PeerIter<'a> = Box<dyn Iterator<Item = &'a Peer> + 'a>;
+pub trait EvictionIterExt<'a> {
+    fn from_peers(peers: &'a [&'a Peer]) -> EvictionIter<'a>;
+
+    fn filter_peers<F>(self, amount: usize, compare_fn: F) -> EvictionIter<'a>
+    where
+        F: Fn(&EvictionRanks, &EvictionRanks) -> Ordering;
+
+    fn select_peers_weighted<F>(self, amount: usize, weight_fn: F) -> EvictionIter<'a>
+    where
+        F: Fn(&EvictionRanks) -> f64;
+
+    fn iterate_peers(self) -> PeerIter<'a>;
+}
```

Comment:
why do u need box here?
trait function can return impl Iterator

Reply:
[my pr to your branch](https://github.com/D-Stacks/rusty-kaspa/pull/6)

Reply:
thanks for this insight, i took the wrong turn to circumvent some compile errors, it seems, and didn't know this about trait functions.. will merge.